### ISSUE TITLE
[spirv] PR #2773 prevents usage of discard without SPV_EXT_demote_to_helper_invocation

### ISSUE THREAD

**Issue by awefers** [microsoft/DirectXShaderCompiler#2823]
PR #2773 (commit f6057fbee2f597a9f343215a98492d04674f1d6b) did add support for SPV_EXT_demote_to_helper_invocation. The side effect of this change is that usage of the discard keyword now requires the availability of the extension VK_EXT_shader_demote_to_helper_invocation. This breaks shaders that are using discard on devices that do not support this extension, like for example Android devices that cannot add retroactively support for this new SPIRV opcode/extension. As a work-around, can the requirement for SPV_EXT_demote_to_helper_invocation be relaxed so that it is optional instead of mandatory? Ie. only emit OpDemoteToHelperInvocationEXT if "-fspv-extension=SPV_EXT_demote_to_helper_invocation" is passed as compiler option. If it it is not specified, emit OpKill like in the past and do not throw the compiler error "error: SPIR-V extension "SPV_EXT_demote_to_helper_invocation" required for discard but not permitted to use"?

**Comment by ehsannas** [comment]
Thanks for bringing this up @awefers. I agree we should somehow make it possible to get the old OpKill behavior.

Do note that using OpKill for `discard` is in fact **_incorrect_**. The HLSL `discard`'s intent is not the same as OpKill.

From the [SPV_EXT_demote_to_helper_invocation spec](https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/EXT/SPV_EXT_demote_to_helper_invocation.asciidoc):
> This extension adds a new OpDemoteToHelperInvocationEXT instruction which is similar to OpKill in that it suppresses subsequent stores and writes to outputs, but is not a flow control instruction and does not necessarily terminate the shader invocation. This is a better match for D3D’s discard instruction, and preserves the ability to rely on uniform flow control for derivatives after the discard.

So we are reluctant to have DXC generate OpKill by default. I'll work on a command line option to get the old behavior, and we'll hopefully be able to clean things up once the extension becomes part of the core Vulkan.

**Comment by awefers** [comment]
What about emitting a warning to indicate the usage is incorrect without SPV_EXT_demote_to_helper_invocation? We are using -WX so we would get initially an error without white-listing the warning. That seems like a good compromise?

**Comment by ehsannas** [comment]
@awefers I've landed the fix, so you should be able to compile `discard` without requiring the extension.

You can also get the "new" behavior (demote) using the 
`-fspv-extension=SPV_EXT_demote_to_helper_invocation` command line option.

Thanks

**Comment by awefers** [comment]
@ehsannas Thank you so much! I already pulled master and it all works as expected!

**Comment by webanck** [comment]
Hello, it seems there has been some regression on the ` OpDemoteToHelperInvocation` matter.

Take the following pixel shader example:
```hlsl
float4 pixelShader(float4 p : SV_POSITION) : SV_TARGET
{
	if (p.x == 0.) discard;
    return float4(0., 0., 0., 0.);
}
```

Compiling it with DXC version `1.7 - 1.7.0.4004 (4cfac0142)`: `.\dxc.exe -E pixelShader -T ps_6_2  -spirv -fspv-target-env="vulkan1.3" .\test.hlsl` outputs the following:
```
; SPIR-V
; Version: 1.6
; Generator: Google spiregg; 0
; Bound: 19
; Schema: 0
               OpCapability Shader
               OpCapability DemoteToHelperInvocation
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %pixelShader "pixelShader" %gl_FragCoord %out_var_SV_TARGET
               OpExecutionMode %pixelShader OriginUpperLeft
               OpSource HLSL 620
               OpName %out_var_SV_TARGET "out.var.SV_TARGET"
               OpName %pixelShader "pixelShader"
               OpDecorate %gl_FragCoord BuiltIn FragCoord
               OpDecorate %out_var_SV_TARGET Location 0
      %float = OpTypeFloat 32
    %float_0 = OpConstant %float 0
    %v4float = OpTypeVector %float 4
          %7 = OpConstantComposite %v4float %float_0 %float_0 %float_0 %float_0
%_ptr_Input_v4float = OpTypePointer Input %v4float
%_ptr_Output_v4float = OpTypePointer Output %v4float
       %void = OpTypeVoid
         %11 = OpTypeFunction %void
       %bool = OpTypeBool
%gl_FragCoord = OpVariable %_ptr_Input_v4float Input
%out_var_SV_TARGET = OpVariable %_ptr_Output_v4float Output
%pixelShader = OpFunction %void None %11
         %13 = OpLabel
         %14 = OpLoad %v4float %gl_FragCoord
         %15 = OpCompositeExtract %float %14 0
         %16 = OpFOrdEqual %bool %15 %float_0
               OpSelectionMerge %17 None
               OpBranchConditional %16 %18 %17
         %18 = OpLabel
               OpDemoteToHelperInvocation
               OpBranch %17
         %17 = OpLabel
               OpStore %out_var_SV_TARGET %7
               OpReturn
               OpFunctionEnd
```

As you can see, it outputs an `OpDemoteToHelperInvocation` operation near the end. 

This doesn't happen with Vulkan target versions 1.0, 1.1, and 1.2; it starts at target 1.3. So, is there any way to prevent those operations to be output for the Vulkan target 1.3?
